lib/http-bridge/test/integration/Test/Integration/Framework/DSL.hs

```diff
@@ -306,25 +312,77 @@ expectEventually
 expectEventually ctx getter target (_, res) = case res of
     Left e -> wantedSuccessButError e
     Right s -> loopUntilRestore (s ^. walletId)
-
   where
     loopUntilRestore :: (MonadIO m, MonadCatch m) => Text -> m ()
     loopUntilRestore wid = do
         r <- request @ApiWallet ctx ("GET", "v2/wallets/" <> wid) Default Empty
         let target' = getFromResponse getter r
         unless (target' >= target) $ loopUntilRestore wid
 
+expectEventually'
+    :: (MonadIO m, MonadCatch m, MonadFail m, Ord a)
+    => Context
+    -> Lens' ApiWallet a
+    -> a
+    -> ApiWallet
+    -> m ()
+expectEventually' ctx target value wallet = do
+    rb <- request @ApiWallet ctx (getWallet wallet) Default Empty
+    expectEventually ctx target value rb
+--
+-- CLI output expectations
+--
+
 -- | Expects a given string to be a valid JSON output corresponding to some
--- given data-type 'a'
+-- given data-type 'a'. Returns this type if successful.
 expectValidJSON
     :: forall m a. (MonadFail m, FromJSON a)
     => Proxy a
     -> String
-    -> m ()
+    -> m a
 expectValidJSON _ str =
     case Aeson.eitherDecode @a (BL8.pack str) of
         Left e -> fail $ "expected valid JSON but failed decoding: " <> show e
-        Right _ -> return ()
+        Right a -> return a
+
+-- | Expects wallet from the request to eventually reach the given state or
+-- beyond
+-- expectCliEventually
+--     :: (MonadIO m, MonadCatch m, MonadFail m, Ord a)
+--     => Lens' ApiWallet a
+--     -> a
+--     -> ApiWallet
+--     -> m ()
+-- expectCliEventually getter target w = loopUntilRestore (w ^. walletId)
+--   where
+--     loopUntilRestore :: (MonadIO m, MonadCatch m) => Text -> IO ()
+--     loopUntilRestore wid = do
+--         Stdout out <- getWalletViaCLI (T.unpack wid)
+--         outJson <- expectValidJSON (Proxy @ApiWallet) out
+--         let target' = view getter outJson
+--         unless (target' >= target) $ loopUntilRestore wid
+
+expectCliFieldBetween
+    :: (MonadIO m, MonadFail m, Show a, Ord a)
+    => Lens' s a
+    -> (a, a)
+    -> s
+    -> m ()
+expectCliFieldBetween getter (aMin, aMax) s = case view getter s of
+            a | a < aMin -> fail $
+                "expected " <> show a <> " >= " <> show aMin
+            a | a > aMax -> fail $
+                "expected " <> show a <> " <= " <> show aMax
+            _ ->
+                return ()
+
+expectCliFieldEqual
+    :: (MonadIO m, MonadFail m, Show a, Eq a)
+    => Lens' s a
+    -> a
+    -> s
+    -> m ()
+expectCliFieldEqual getter a out = (view getter out) `shouldBe` a
```

Comment:
Hmmm... There start to be some obvious duplication in the DSL we could most probably get rid of by some clever tricks with the type-system and lenses here. I'll give it a shot.

Reply:
:+1: